src/geom/util/GeometryFixer.cpp

```diff
@@ -0,0 +1,337 @@
+
+    auto ptsFix = RepeatedPointRemover::removeRepeatedAndInvalidPoints(p_geom->getCoordinatesRO());
+    auto ptsFixSz = ptsFix->size();
+    if (isKeepCollapsed && ptsFixSz == 1) {
+        const Coordinate& c = ptsFix->getAt(0);
+        std::unique_ptr<Geometry> pt(factory->createPoint(c));
+        return pt;
+    }
+    if (ptsFixSz <= 1)
+        return std::unique_ptr<Geometry>(nullptr);
+    else
+        return factory->createLineString(std::move(ptsFix));
+}
+
+
+/* private */
+std::unique_ptr<Geometry>
+GeometryFixer::fixMultiLineString(const MultiLineString* p_geom)
+{
+    std::vector<std::unique_ptr<Geometry>> fixed;
+    bool isMixed = false;
+    for (std::size_t i = 0; i < p_geom->getNumGeometries(); i++) {
+        const LineString* line = p_geom->getGeometryN(i);
+        if (line->isEmpty())
+            continue;
+
+        std::unique_ptr<Geometry> fix = fixLineStringElement(line);
+        if (fix == nullptr)
+            continue;
+
+        if (! (fix->getGeometryTypeId() == GEOS_LINESTRING)) {
+            isMixed = true;
+        }
+        fixed.emplace_back(fix.release());
+    }
+    if (fixed.size() == 1) {
+        std::unique_ptr<Geometry> g(fixed.at(0).release());
+        return g;
+    }
+    if (isMixed)
+        return factory->createGeometryCollection(std::move(fixed));
+    else
+        return factory->createMultiLineString(std::move(fixed));
+}
+
+/* private */
+std::unique_ptr<Geometry>
+GeometryFixer::fixPolygon(const Polygon* p_geom)
+{
+    std::unique_ptr<Geometry> fix = fixPolygonElement(p_geom);
+    if (fix == nullptr)
+        return factory->createPolygon();
+    else
+        return fix;
+}
+
+/* private */
+std::unique_ptr<Geometry>
+GeometryFixer::fixPolygonElement(const Polygon* p_geom)
+{
+    const LinearRing* shell = p_geom->getExteriorRing();
+    std::unique_ptr<Geometry> fixShell = fixRing(shell);
+    if (fixShell->isEmpty()) {
+        if (isKeepCollapsed) {
+            const CoordinateSequence* cs = shell->getCoordinatesRO();
+            std::unique_ptr<LineString> line(factory->createLineString(*cs));
+            return fixLineString(line.get());
+        }
+        //-- if not allowing collapses then return empty polygon
+        return std::unique_ptr<Geometry>(nullptr);
+    }
+    // if no holes then done
+    if (p_geom->getNumInteriorRing() == 0) {
+        return fixShell;
+    }
+    std::unique_ptr<Geometry> fixedHoles = fixHoles(p_geom);
+    std::unique_ptr<Geometry> result = removeHoles(fixShell.get(), fixedHoles.get());
+    return result;
+}
+
+/* private */
+std::unique_ptr<Geometry>
+GeometryFixer::fixHoles(const Polygon* p_geom)
+{
+    std::vector<std::unique_ptr<Geometry>> holes;
+
+    for (std::size_t i = 0; i < p_geom->getNumInteriorRing(); i++) {
+        std::unique_ptr<Geometry> holeRep = fixRing(p_geom->getInteriorRingN(i));
+        // Do not preserve null/empty holes
+        if (!(holeRep == nullptr || holeRep->isEmpty())) {
```

Comment:
```suggestion
        if (holeRep != nullptr && ! holeRep->isEmpty()) {
```
at least for consistency with the style used in GeometryFixer::fixMultiPolygon(